### PR TITLE
Provide ability to retrieve original Request object from Call.

### LINKED ISSUE #1367: Provide new method to get the original request from Retrofit Call proxies
Hi,

When using Retrofit Call proxies, it's not possible to be aware of the original request through those objects. When implementing a custom response cache (not depending on response cache of okhttp), generating a cache key with the request is necessary. However we cannot access to this original request, headers, params through Call objects (okhttpcall being final as well).

Would be nice to provide in the interface Call the method CreateRawCall already implemented in private.

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #1367. Closes #1369.

### CODE REVIEW

File: retrofit/src/test/java/retrofit2/CallTest.java [hunk 1]
```diff
@@ -667,4 +670,269 @@
...
+    Object a = new Object() {
+      @Override public String toString() {
+        writeCount.incrementAndGet();
+        throw new RuntimeException("Broken!");
+      }
+    };
+    Call<String> call = service.postRequestBody(a);
+
+    try {
+      call.request();
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit2/CallTest.java]: fail()

File: retrofit/src/test/java/retrofit2/CallTest.java [hunk 2]
```diff
@@ -667,4 +670,269 @@
...
+    } catch (RuntimeException e) {
+      assertThat(e).hasMessage("Broken!");
+    }
+    assertThat(writeCount.get()).isEqualTo(1);
+
+    try {
+      call.execute();
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit2/CallTest.java]: fail()

File: retrofit/src/test/java/retrofit2/CallTest.java [hunk 3]
```diff
@@ -667,4 +670,269 @@
...
+    Object a = new Object() {
+      @Override public String toString() {
+        writeCount.incrementAndGet();
+        throw new RuntimeException("Broken!");
+      }
+    };
+    Call<String> call = service.postRequestBody(a);
+
+    try {
+      call.execute();
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit2/CallTest.java]: fail()

File: retrofit/src/test/java/retrofit2/CallTest.java [hunk 4]
```diff
@@ -667,4 +670,269 @@
...
+    } catch (RuntimeException e) {
+      assertThat(e).hasMessage("Broken!");
+    }
+    assertThat(writeCount.get()).isEqualTo(1);
+
+    try {
+      call.request();
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit2/CallTest.java]: fail()

File: retrofit/src/test/java/retrofit2/CallTest.java [hunk 5]
```diff
@@ -667,4 +670,269 @@
...
+    Object a = new Object() {
+      @Override public String toString() {
+        writeCount.incrementAndGet();
+        throw new RuntimeException("Broken!");
+      }
+    };
+    Call<String> call = service.postRequestBody(a);
+
+    try {
+      call.request();
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit2/CallTest.java]: fail()

File: retrofit/src/test/java/retrofit2/CallTest.java [hunk 6]
```diff
@@ -667,4 +670,269 @@
...
+      @Override public String toString() {
+        writeCount.incrementAndGet();
+        throw new RuntimeException("Broken!");
+      }
+    };
+    Call<String> call = service.postRequestBody(a);
+
+    final CountDownLatch latch = new CountDownLatch(1);
+    call.enqueue(new Callback<String>() {
+      @Override public void onResponse(Call<String> call, Response<String> response) {
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit2/CallTest.java]: throw new AssertionError ?

File: retrofit/src/test/java/retrofit2/CallTest.java [hunk 7]
```diff
@@ -667,4 +670,269 @@
...
+      @Override public void onFailure(Call<String> call, Throwable t) {
+        assertThat(t).isExactlyInstanceOf(RuntimeException.class).hasMessage("Broken!");
+        assertThat(writeCount.get()).isEqualTo(1);
+        latch.countDown();
+      }
+    });
+    assertTrue(latch.await(1, SECONDS));
+
+    try {
+      call.request();
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit2/CallTest.java]: fail()

**Comment by swankjesse** [comment]
LGTM

File: retrofit/src/test/java/retrofit2/CallTest.java (hunk 6 above)
Comment (JakeWharton) [inline retrofit/src/test/java/retrofit2/CallTest.java]: The latch will just timeout if the wrong thing is called. If you throw here you'll get double stack traces so I left it out.

**Comment by JakeWharton** [comment]
`fail()`s added! Forget it once, copy/paste it 12 times...

**Pull request merged**